stl/src/vector_algorithms.cpp

```diff
@@ -3384,30 +3881,59 @@ namespace {
                     }
                 }
 
-                const int _Last_part_size_el = static_cast<int>(_Last_part_size / sizeof(_Ty));
-                const __m128i _Data1         = _mm_loadu_si128(reinterpret_cast<const __m128i*>(_Haystack));
-                _Test_whole_needle(_Data1, _Last_part_size_el);
+                if (_Last_part_size != 0) {
+                    const int _Last_part_size_el = static_cast<int>(_Last_part_size / sizeof(_Ty));
+                    __m128i _Data1;
+
+                    if (_Haystack_length_bytes >= 16) {
+                        _Data1 = _mm_loadu_si128(reinterpret_cast<const __m128i*>(_Haystack));
+                    } else {
+                        alignas(16) uint8_t _Tmp1[16];
+                        memcpy(_Tmp1, _Haystack, _Haystack_length_bytes);
+                        _Data1 = _mm_load_si128(reinterpret_cast<const __m128i*>(_Tmp1));
+                    }
+
+                    _Test_whole_needle(_Data1, _Last_part_size_el);
+                }
 
                 return static_cast<size_t>(_Found_pos);
             }
         }
 #endif // !_M_ARM64EC
-        const auto _Ptr_haystack = static_cast<const _Ty*>(_Haystack);
-        size_t _Pos              = _Haystack_length;
-        const auto _Needle_end   = static_cast<const _Ty*>(_Needle) + _Needle_length;
 
-        while (_Pos != 0) {
-            --_Pos;
+        template <class _Ty>
+        size_t _Dispatch_pos(const void* const _Haystack, const size_t _Haystack_length, const void* const _Needle,
+            const size_t _Needle_length) noexcept {
+#ifndef _M_ARM64EC
+            if (_Use_sse42()) {
+                if (_Use_avx2()) {
+                    if (_Bitmap::_Use_bitmap_avx<_Ty>(_Haystack_length, _Needle_length)
+                        && _Bitmap::_Can_fit_256_bits_sse(static_cast<const _Ty*>(_Needle), _Needle_length)) {
+                        return _Bitmap::_Impl_last_avx<_Ty>(_Haystack, _Haystack_length, _Needle, _Needle_length);
+                    }
+                } else {
+                    if (_Bitmap::_Use_bitmap_sse<_Ty>(_Haystack_length, _Needle_length)
+                        && _Bitmap::_Can_fit_256_bits_sse(static_cast<const _Ty*>(_Needle), _Needle_length)) {
+                        _Bitmap::_Scalar_table_t _Table = {};
+                        _Bitmap::_Build_scalar_table_no_check<_Ty>(_Needle, _Needle_length, _Table);
+                        return _Bitmap::_Impl_last_scalar<_Ty>(_Haystack, _Haystack_length, _Table);
+                    }
+                }

```

Comment:
Question about this control flow: If we can use AVX2, but the `_Use_bitmap_avx` threshold isn't met, we don't try the `_Use_bitmap_sse` threshold, and instead just fall through. But if we can only use SSE4.2, we try the SSE4.2 threshold before falling through, indicating that the guarded code is valuable. From a brief glance, I see that there are cases where the AVX2 threshold wouldn't activate, but the SSE4.2 threshold would.

Given that the threshold checks are cheap integer comparisons, should we follow the more common pattern of "if we can use AVX2 and the AVX2 threshold is met, return cool AVX2 thing. If we can use SSE4.2 and the SSE4.2 threshold is met, return slightly less cool SSE4.2 thing. Return fallback thing."?

Reply:
I wanted to approximate the problem by not looking into these cases and not picking between three algorithms.

The cases we pessimize by this approximation are the cases where the needle greatly exceeds the haystack, and the pessimization is by some noticeable percentage, but not like 2x.

----

If we want to do this right, we should handle not only the cases you see here: "AVX bitmap is not preferable over vector nested loop, but scalar bitmap is preferable over vector nested loop", but also a set of cases "AVX bitmap is preferable over vector nested loop, and scalar bitmap is also preferable over vector nested loop, and scalar bitmap is preferable over AVX bitmap". The latter is a greater set.

At some point the AVX threshold function returned enum, we need to return to that:
```C++
enum class _Strategy { _No_bitmap, _Scalar_bitmap, _Vector_bitmap };

_Strategy _Pick_strategy_avx(const size_t _Count1, const size_t _Count2) noexcept { ... }
```

----

So I don't mind handling these cases better, at the expense of threshold complication, do you want me to do that?